File: sdk/go/examples/basic_example.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	kvdb "github.com/kvdb/go-client"
)

func main() {
	// 配置客户端
	config := kvdb.DefaultClientConfig()
	config.ServerAddress = "localhost:50051"
	config.ConnectionTimeout = 5 * time.Second
	config.RequestTimeout = 10 * time.Second
	
	// 创建客户端
	client := kvdb.NewClient(config)
	defer client.Close()
	
	ctx := context.Background()
	
	// 连接到服务器
	if err := client.Connect(ctx); err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	
	fmt.Println("Connected to KVDB server")
	
	// 基本操作示例
	if err := basicOperationsExample(ctx, client); err != nil {
		log.Printf("Basic operations error: %v", err)
	}
	
	// 批量操作示例
	if err := batchOperationsExample(ctx, client); err != nil {
		log.Printf("Batch operations error: %v", err)
	}
	
	// 扫描操作示例
	if err := scanOperationsExample(ctx, client); err != nil {
		log.Printf("Scan operations error: %v", err)
	}
	
	// 快照操作示例
	if err := snapshotOperationsExample(ctx, client); err != nil {
		log.Printf("Snapshot operations error: %v", err)
	}
	
	// 订阅示例
	if err := subscriptionExample(ctx, client); err != nil {
		log.Printf("Subscription error: %v", err)
	}
	
	// 管理操作示例
	if err := managementOperationsExample(ctx, client); err != nil {
		log.Printf("Management operations error: %v", err)
	}
}

func basicOperationsExample(ctx context.Context, client *kvdb.Client) error {
	fmt.Println("\n=== Basic Operations Example ===")
	
	// PUT操作
	if err := client.Put(ctx, "user:1001", "Alice"); err != nil {
		return fmt.Errorf("PUT failed: %w", err)
	}
	fmt.Println("PUT user:1001 = Alice")
	
	// GET操作
	value, found, err := client.Get(ctx, "user:1001")
	if err != nil {
		return fmt.Errorf("GET failed: %w", err)
	}
	if found {
		fmt.Printf("GET user:1001 = %s\n", value)
	} else {
		fmt.Println("GET user:1001 = <not found>")
	}
	
	// DELETE操作
	if err := client.Delete(ctx, "user:1001"); err != nil {
		return fmt.Errorf("DELETE failed: %w", err)
	}
	fmt.Println("DELETE user:1001")
	
	// 验证删除
	value, found, err = client.Get(ctx, "user:1001")
	if err != nil {
		return fmt.Errorf("GET after delete failed: %w", err)
	}
	if found {
		fmt.Printf("GET user:1001 after delete = %s\n", value)
	} else {
		fmt.Println("GET user:1001 after delete = <not found>")
	}
	
	return nil
}

func batchOperationsExample(ctx context.Context, client *kvdb.Client) error {
	fmt.Println("\n=== Batch Operations Example ===")
	
	// 批量PUT
	pairs := []kvdb.KeyValue{
		{Key: "user:1001", Value: "Alice"},
		{Key: "user:1002", Value: "Bob"},
		{Key: "user:1003", Value: "Charlie"},
		{Key: "user:1004", Value: "David"},
	}
	
	if err := client.BatchPut(ctx, pairs); err != nil {
		return fmt.Errorf("batch PUT failed: %w", err)
	}
	fmt.Printf("Batch PUT %d pairs\n", len(pairs))
	
	// 批量GET
	keys := []string{"user:1001", "user:1002", "user:1003"}
	results, err := client.BatchGet(ctx, keys)
	if err != nil {
		return fmt.Errorf("batch GET failed: %w", err)
	}
	
	fmt.Println("Batch GET results:")
	for _, kv := range results {
		fmt.Printf("  %s = %s\n", kv.Key, kv.Value)
	}
	
	return nil
}

func scanOperationsExample(ctx context.Context, client *kvdb.Client) error {
	fmt.Println("\n=== Scan Operations Example ===")
	
	// 前缀扫描
	results, err := client.PrefixScan(ctx, "user:", 10)
	if err != nil {
		return fmt.Errorf("prefix scan failed: %w", err)
	}
	
	fmt.Println("Prefix scan results:")
	for _, kv := range results {
		fmt.Printf("  %s = %s\n", kv.Key, kv.Value)
	}
	
	// 范围扫描
	scanOptions := &kvdb.ScanOptions{
		StartKey: "user:1000",
		EndKey:   "user:1999",
		Limit:    5,
	}
	
	results, err = client.Scan(ctx, scanOptions)
	if err != nil {
		return fmt.Errorf("range scan failed: %w", err)
	}
	
	fmt.Println("Range scan results:")
	for _, kv := range results {
		fmt.Printf("  %s = %s\n", kv.Key, kv.Value)
	}
	
	return nil
}

func snapshotOperationsExample(ctx context.Context, client *kvdb.Client) error {
	fmt.Println("\n=== Snapshot Operations Example ===")
	
	// 创建快照
	snapshot, err := client.CreateSnapshot(ctx)
	if err != nil {
		return fmt.Errorf("create snapshot failed: %w", err)
	}
	fmt.Printf("Created snapshot: %d\n", snapshot.ID)
	
	// 在快照上读取数据
	value, found, err := client.GetAtSnapshot(ctx, "user:1001", snapshot)
	if err != nil {
		return fmt.Errorf("get at snapshot failed: %w", err)
	}
	if found {
		fmt.Printf("Snapshot GET user:1001 = %s\n", value)
	} else {
		fmt.Println("Snapshot GET user:1001 = <not found>")
	}
	
	// 修改数据
	if err := client.Put(ctx, "user:1001", "Alice_Modified"); err != nil {
		return fmt.Errorf("PUT failed: %w", err)
	}
	
	// 当前读取
	currentValue, found, err := client.Get(ctx, "user:1001")
	if err != nil {
		return fmt.Errorf("current GET failed: %w", err)
	}
	if found {
		fmt.Printf("Current GET user:1001 = %s\n", currentValue)
	}
	
	// 快照读取（应该是旧值）
	snapshotValue, found, err := client.GetAtSnapshot(ctx, "user:1001", snapshot)
	if err != nil {
		return fmt.Errorf("snapshot GET failed: %w", err)
	}
	if found {
		fmt.Printf("Snapshot GET user:1001 = %s\n", snapshotValue)
	}
	
	// 释放快照
	if err := client.ReleaseSnapshot(ctx, snapshot); err != nil {
		return fmt.Errorf("release snapshot failed: %w", err)
	}
	fmt.Println("Released snapshot")
	
	return nil
}

func subscriptionExample(ctx context.Context, client *kvdb.Client) error {
	fmt.Println("\n=== Subscription Example ===")
	
	// 开始订阅
	subscription, err := client.Subscribe(ctx, "user:*", func(event *kvdb.SubscriptionEvent) {
		fmt.Printf("Subscription event: %s %s = %s\n", event.Operation, event.Key, event.Value)
	}, true)
	
	if err != nil {
		return fmt.Errorf("subscribe failed: %w", err)
	}
	
	fmt.Println("Started subscription")
	
	// 等待一段时间
	time.Sleep(1 * time.Second)
	
	// 触发一些事件
	client.Put(ctx, "user:1005", "Eve")
	client.Put(ctx, "user:1006", "Frank")
	client.Delete(ctx, "user:1004")
	
	// 等待事件处理
	time.Sleep(2 * time.Second)
	
	// 取消订阅
	subscription.Cancel()
	fmt.Println("Cancelled subscription")
	
	return nil
}

func managementOperationsExample(ctx context.Context, client *kvdb.Client) error {
	fmt.Println("\n=== Management Operations Example ===")
	
	// 获取统计信息
	stats, err := client.GetStats(ctx)
	if err != nil {
		return fmt.Errorf("get stats failed: %w", err)
	}
	
	fmt.Println("Database stats:")
	fmt.Printf("  Memtable size: %d\n", stats.MemtableSize)
	fmt.Printf("  WAL size: %d\n", stats.WALSize)
	fmt.Printf("  Cache hit rate: %.2f\n", stats.CacheHitRate)
	fmt.Printf("  Active snapshots: %d\n", stats.ActiveSnapshots)
	
	// 刷新数据
	if err := client.Flush(ctx); err != nil {
		return fmt.Errorf("flush failed: %w", err)
	}
	fmt.Println("Flushed data to disk")
	
	// 压缩数据
	if err := client.Compact(ctx); err != nil {
		return fmt.Errorf("compact failed: %w", err)
	}
	fmt.Println("Compacted data")
	
	return nil
}

File: sdk/python/examples/basic_example.py
#!/usr/bin/env python3
"""
KVDB Python客户端基础示例
"""

import asyncio
import time
from kvdb_client import KVDBClient, ClientConfig, KeyValue, ScanOptions


def basic_operations_example():
    """基本操作示例"""
    print("=== Basic Operations Example ===")
    
    # 配置客户端
    config = ClientConfig(
        server_address="localhost:50051",
        protocol="grpc",
        connection_timeout=5.0,
        request_timeout=10.0
    )
    
    # 使用上下文管理器自动连接和断开
    with KVDBClient(config) as client:
        print("Connected to KVDB server")
        
        # PUT操作
        client.put("user:1001", "Alice")
        print("PUT user:1001 = Alice")
        
        # GET操作
        value = client.get("user:1001")
        print(f"GET user:1001 = {value}")
        
        # DELETE操作
        client.delete("user:1001")
        print("DELETE user:1001")
        
        # 验证删除
        value = client.get("user:1001")
        print(f"GET user:1001 after delete = {value}")


def batch_operations_example():
    """批量操作示例"""
    print("\n=== Batch Operations Example ===")
    
    config = ClientConfig(server_address="localhost:50051")
    
    with KVDBClient(config) as client:
        # 批量PUT
        pairs = [
            KeyValue("user:1001", "Alice"),
            KeyValue("user:1002", "Bob"),
            KeyValue("user:1003", "Charlie"),
            KeyValue("user:1004", "David")
        ]
        
        client.batch_put(pairs)
        print(f"Batch PUT {len(pairs)} pairs")
        
        # 批量GET
        keys = ["user:1001", "user:1002", "user:1003"]
        results = client.batch_get(keys)
        
        print("Batch GET results:")
        for kv in results:
            print(f"  {kv.key} = {kv.value}")


def scan_operations_example():
    """扫描操作示例"""
    print("\n=== Scan Operations Example ===")
    
    config = ClientConfig(server_address="localhost:50051")
    
    with KVDBClient(config) as client:
        # 前缀扫描
        results = client.prefix_scan("user:", limit=10)
        print("Prefix scan results:")
        for kv in results:
            print(f"  {kv.key} = {kv.value}")
        
        # 范围扫描
        scan_options = ScanOptions(
            start_key="user:1000",
            end_key="user:1999",
            limit=5
        )
        
        results = client.scan(scan_options)
        print("Range scan results:")
        for kv in results:
            print(f"  {kv.key} = {kv.value}")


def snapshot_operations_example():
    """快照操作示例"""
    print("\n=== Snapshot Operations Example ===")
    
    config = ClientConfig(server_address="localhost:50051")
    
    with KVDBClient(config) as client:
        # 创建快照
        snapshot = client.create_snapshot()
        print(f"Created snapshot: {snapshot}")
        
        # 在快照上读取数据
        value = client.get_at_snapshot("user:1001", snapshot)
        print(f"Snapshot GET user:1001 = {value}")
        
        # 修改数据
        client.put("user:1001", "Alice_Modified")
        
        # 当前读取
        current_value = client.get("user:1001")
        print(f"Current GET user:1001 = {current_value}")
        
        # 快照读取（应该是旧值）
        snapshot_value = client.get_at_snapshot("user:1001", snapshot)
        print(f"Snapshot GET user:1001 = {snapshot_value}")
        
        # 释放快照
        client.release_snapshot(snapshot)
        print("Released snapshot")


async def async_operations_example():
    """异步操作示例"""
    print("\n=== Async Operations Example ===")
    
    config = ClientConfig(server_address="localhost:50051")
    
    with KVDBClient(config) as client:
        # 异步PUT操作
        tasks = []
        for i in range(5):
            task = client.put_async(f"async:key{i}", f"value{i}")
            tasks.append(task)
        
        # 等待所有PUT操作完成
        results = await asyncio.gather(*tasks)
        print(f"Completed {len(results)} async PUT operations")
        
        # 异步GET操作
        get_tasks = []
        for i in range(5):
            task = client.get_async(f"async:key{i}")
            get_tasks.append(task)
        
        # 等待所有GET操作完成
        values = await asyncio.gather(*get_tasks)
        print("Async GET results:")
        for i, value in enumerate(values):
            print(f"  async:key{i} = {value}")


def subscription_example():
    """订阅示例"""
    print("\n=== Subscription Example ===")
    
    config = ClientConfig(server_address="localhost:50051")
    
    with KVDBClient(config) as client:
        # 定义回调函数
        def on_change(key, value, operation):
            print(f"Subscription event: {operation} {key} = {value}")
        
        # 开始订阅
        subscription_id = client.subscribe("user:*", on_change, include_deletes=True)
        print(f"Started subscription {subscription_id}")
        
        # 等待一段时间
        time.sleep(1)
        
        # 触发一些事件
        client.put("user:1005", "Eve")
        client.put("user:1006", "Frank")
        client.delete("user:1004")
        
        # 等待事件处理
        time.sleep(2)
        
        # 取消订阅
        client.cancel_subscription(subscription_id)
        print("Cancelled subscription")


def management_operations_example():
    """管理操作示例"""
    print("\n=== Management Operations Example ===")
    
    config = ClientConfig(server_address="localhost:50051")
    
    with KVDBClient(config) as client:
        # 获取统计信息
        stats = client.get_stats()
        print("Database stats:")
        for key, value in stats.items():
            print(f"  {key} = {value}")
        
        # 刷新数据
        client.flush()
        print("Flushed data to disk")
        
        # 压缩数据
        client.compact()
        print("Compacted data")


def http_client_example():
    """HTTP客户端示例"""
    print("\n=== HTTP Client Example ===")
    
    config = ClientConfig(
        server_address="localhost:50051",
        protocol="http",
        http_base_url="http://localhost:8080"
    )
    
    try:
        with KVDBClient(config) as client:
            # 基本操作
            client.put("http:test", "HTTP Value")
            value = client.get("http:test")
            print(f"HTTP GET result: {value}")
            
            # 批量操作
            pairs = [KeyValue("http:1", "value1"), KeyValue("http:2", "value2")]
            client.batch_put(pairs)
            
            results = client.batch_get(["http:1", "http:2"])
            print("HTTP batch results:")
            for kv in results:
                print(f"  {kv.key} = {kv.value}")
                
    except Exception as e:
        print(f"HTTP client error (server may not be running): {e}")


def main():
    """主函数"""
    try:
        basic_operations_example()
        batch_operations_example()
        scan_operations_example()
        snapshot_operations_example()
        
        # 异步操作需要事件循环
        asyncio.run(async_operations_example())
        
        subscription_example()
        management_operations_example()
        http_client_example()
        
    except Exception as e:
        print(f"Error: {e}")
        print("Make sure KVDB server is running on localhost:50051")


if __name__ == "__main__":
    main()